### ISSUE TITLE
Internal DeadLock exception when creating Indexes

### ISSUE THREAD

**Issue by neoecos** [neo4j/neo4j#1878]
Short: Index creation on non-created labels produces a DeadLockException

Large:

Hi, i've been using neo4j the last year, i use py2neo and neomodel, now i'm trying to migrate to neo4j 2.0 and py2neo 1.6.1 neomodel 1.0.

The first thing neomodel tries to do is to create index on properties, even without ever create the label or any node, and when it does that it executes a query like the followings, also y tried executing the same queries on the neo4j browser and i always got the same exception

```
CREATE CONSTRAINT on (n:CaliopeNode) ASSERT n.uuid IS UNIQUE;
CREATE INDEX on :VersionedNode(change_info);
```

```
CYPHER
CREATE INDEX on :VersionedNode(change_info);

Don't panic.

A deadlock scenario has been detected and avoided. This means that two or more transactions, which were holding locks, were wanting to await locks held by one another, which would have resulted in a deadlock between these transactions. This exception was thrown instead of ending up in that deadlock.

See the deadlock section in the Neo4j manual for how to avoid this: http://docs.neo4j.org/chunked/stable/transactions-deadlocks.html

Details: 'Transaction(203, owner:"qtp1985350784-315")[STATUS_ACTIVE,Resources=1] can't wait on resource RWLock[SchemaLock] since => Transaction(203, owner:"qtp1985350784-315")[STATUS_ACTIVE,Resources=1] <-[:HELD_BY]- RWLock[SchemaLock] <-[:WAITING_FOR]- Transaction(152, owner:"qtp1985350784-277")[STATUS_ACTIVE,Resources=1] <-[:HELD_BY]- RWLock[SchemaLock]'.

```

Thinking a little bit about the issue, i thought it may be because the labels VersionedNode or CalipeNode never where created, because there is not nodes yet, so i tried to create the label, and it worked.

```
CREATE (n:VersionedNode);
```

After that, tried to create the same index again, and worked. Thank you for reading.

PD: I'm very interested on do a meetup in Bogotá Colombia ! Tell some one who cares!

**Comment by jakewins** [comment]
As this is resolved in the py2neo ticket, closing this.

**Comment by joslinm** [comment]
Hi I'm seeing this in Neo4j 3.0.3.

The exact sequence of events is as follows:
- Neo4j server is doing stuff (this is important -- it can't be idle)
- Try to create an index while it's doing stuff -- it just sits there and hangs. 
- Process stops responding
- Neo4j dies
- Start neo4j again
- Try to create index on same index you did in step 1. 
- Get something like this bacK:

```
DeadlockDetectedException: LockClient[6065] can't wait on resource RWLock[SCHEMA(0), hash=1153058837] since => LockClient[6065] <-[:HELD_BY]- RWLock[SCHEMA(0), hash=1153058837] <-[:WAITING_FOR]- LockClient[6063] <-[:HELD_BY]- RWLock[SCHEMA(0), hash=1153058837]
```

This is returned consistently -- no roll backs or anything happen. I can't create the intended index at all.

**UPDATE**:
Turns out I could create the index using `index --create ...`, but I still get the deadlock error from the Cypher style `CREATE INDEX ON ...` every time even after its creation.

Also, I was wrong about "create same index." I can't create any index without getting the deadlock error back even through Neo4j reboots. 

**UPDATE 2**
I had to resort to creating indexes using `index --create` from shell. `CREATE INDEX ON` times out every time. Yet even doing this appears to time out the Neo4j server. I went to go shut down server and got these ugly messages.

```
[ec2-user@neo4j neo4j-community-3.0.3]$ sudo bin/neo4j console
Starting Neo4j.
2016-06-22 15:49:24.888+0000 INFO  Starting...
2016-06-22 15:49:25.843+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2016-06-22 15:49:29.923+0000 INFO  Started.
2016-06-22 15:49:31.724+0000 INFO  Remote interface available at http://0.0.0.0:7474/
^C2016-06-22 15:54:36.793+0000 INFO  Neo4j Server shutdown initiated by request
2016-06-22 15:54:36.800+0000 WARN  HttpConnection@28552d13{FILLING} s=DISPATCHED i=true a=null
java.lang.IllegalStateException: s=DISPATCHED i=true a=null
    at org.eclipse.jetty.server.HttpChannelState.handling(HttpChannelState.java:232)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:282)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
    at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
    at java.lang.Thread.run(Thread.java:745)
2016-06-22 15:54:36.802+0000 WARN  HttpConnection@7920179a{FILLING} s=DISPATCHED i=true a=null
java.lang.IllegalStateException: s=DISPATCHED i=true a=null
    at org.eclipse.jetty.server.HttpChannelState.handling(HttpChannelState.java:232)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:282)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
    at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
    at java.lang.Thread.run(Thread.java:745)
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.835+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.836+0000 WARN  {} Couldn't stop {}
2016-06-22 15:55:06.842+0000 INFO  Stopping...
2016-06-22 15:55:16.914+0000 INFO  Stopped.
```

**UPDATE 3**
I'm seeing constant timeouts, and believe it's in conjunction with a very long Cypher query being sent. The server won't stay up for more than ~10m without completely timing out and having to get restarted. The timeout does not occur from neo4j shell, but only when accessing it via Cypher endpoint remotely.

**UPDATE 4** _(final?)_
I have isolated this behavior down to a very long Cypher query. I am going to create a new issue about it. It kills both 3.0.3 and the latest 3.1. Port 7474 begins timing out afterwards.

**Comment by chrisvest** [comment]
Hi @joslinm,
I'm not sure it's the same problem. Can I trouble you to open a separate issue (feel free to cross-link) for what you are seeing?

**Comment by jiangnanboy** [comment]
"LockClient[13640] can't wait on resource RWLock[NODE(281315),
